fixes #67 DialPipe problem with multiple calls / waiting for busy pipe

gdamore opened the pull request:
(I believe this change also should be used instead of PR #75 -- that PR is potentially extraordinarily buggy.)

This changes a few things to try to ensure that we never wind up with
a result of ERROR_FILE_NOT_FOUND, due to a race between closing the
last pipe instance and opening the next.

First we keep an "open" client instance (unused) while the listener
is open, so that we are guaranteed to always have an active pipe
instance.  This means attempts to open while no other instances exist
result in ERROR_PIPE_BUSY instead of ERROR_FILE_NOT_FOUND.

Second we have changed the loop for dialing to eliminate a race condition
that is more or less inherent in WaitNamedPipe when synchronizing with
CreateFile.  The real timeout needs to be some larger value than the
WaitNamedPipe timeout, and furthermore WaitNamedPipe is not very nice
with the Go runtime, since it is a blocking system call.  Instead we
just put the goroutine to sleep for 10 milliseconds, and keep retrying
the CreateFile until the maximum timeout is reached.  If no timeout is
specified we assume a reasonable and large default of 5 seconds, which is
similar to a TCP connection timeout.

This isn't perfect, as a client attempting to connect to an extremely
busy pipe server can be starved out by other clients coming in while
it is in that brief sleep, but this potential race was already present
with WaitNamedPipe.  The numerous retries (by default 500 retries!)
mean its pretty unlikely to occur, and if a single client hits the
race once, it has an excellent chance of getting in the next cycle.

(A real "fix" that is completely race free and fair would require
changes in the underlying Named Pipe implementation, or some other
kind of external coordination.)

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/go-winio?pullRequest=80) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/go-winio?pullRequest=80) before we can accept your contribution.<br/><br/>:x: gdamore [sign now](https://cla.opensource.microsoft.com/Microsoft/go-winio?pullRequest=80)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/go-winio?pullRequest=80) it.</sub>

gdamore commented:
Hmm... I probably would have liked to have added my own copyright and update to the code... (still MIT licensed.) If its possible to do that somewhere, text more or less to the effect of:

Portions Copyright 2018 Garrett D'Amore <garrett@damore.org>

(or skip the Portions word.) . If this is too much a hassle, then don't worry about it.

gdamore commented:
This probably also fixes #46 -- since WaitNamedPipe was the only blocking code remaining.

carlfischer1 commented:
cc @jstarks @johnstep

olljanat commented:
FYI. I was able to reproduce original issue by running my test version of Portainer on Windows Server, version 1803 and I cannot see it anymore on version which contains content from this PR (more detail on that portainer PR).

Any change to get this one merged?

StefanScherer commented:
I can confirm that this PR helps Portainer work stable with the named pipe bind mounted into a Windows container running a Docker swarm on Windows Server 1803 with Docker EE 18.03.1-ee-1

gdamore commented:
@jstarks @johnstep Ping?  This PR addresses a very important reliability issue, and at least three different projects are waiting for this to be integrated.  If there is something wrong with this, or concerns with what I've done here, then feedback would very much be appreciated.

johnstep commented:
I do not have permission to merge this.

aiminickwong commented:
@jstarks   time to allow merge this pull ?

gdamore commented:
Hard to believe I submitted this 19 days ago; the silence from the repo maintainers is deafening.

jstarks commented:
Sorry for the delays. I'm looking at this now. I want to write a quick test to understand the behavior of something before I merge this.

jstarks commented:
OK, I think this pull request is actually three changes:
1. Increase the default timeout to 5 seconds.
2. Poll instead of using WaitNamedPipe.
3. Keep the original dummy client handle open for the lifetime of the pipe server.

I understand why increasing the timeout may be useful, but this seems like something the client can easily do without a change to go-winio. I do suspect that the default 50ms is probably too short to be useful given the problems with fairness. But I'm worried that extending it to a full 5 seconds might affect existing clients negatively. Maybe a compromise would be something like 250ms.

WaitNamedPipe has its problems. As you mention, it blocks the OS thread (which go-winio generally tries to avoid), it has various races, it does not guarantee that the client will win the race and be able to connect to the pipe, and most importantly for #67, it does not work reliably inside containers. It does have the advantage, though, that it wakes up immediately when a connection is available. I wonder if a better change wouldn't be to keep the WaitNamedPipe call, but if it fails any reason other than timeout, to sleep for 10ms and loop around again. That would probably resolve #67 and other race conditions.

This suggestion doesn't fix the issue that WaitNamedPipe blocks the OS thread, of course. If this is important to resolve, I would suggest making a change to use `FSCTL_PIPE_WAIT` with an asynchronous DeviceIoControl call. This fsctl is what WaitNamedPipe uses, and it is documented as part of the SMB protocol. It appears to have a non-blocking implementation.

I don't understand the third change. I can't see any behavior differences whether I keep the original client handle open or not. Unless you are aware of a behavior that I am not, I think it's better to close that client handle to avoid unnecessary resource consumption.

olljanat commented:
@jstarks thanks for good comments.

I changed default timeout to 250 ms and added client handle closing now on this commit: https://github.com/olljanat/go-winio/commit/972aaec17501edc2ae66b43541f799eec50cf7c5 and did some testing with Portainer and I can tell that this combination at least still fixes #67

@gdamore I think that it is better if you comment about suggestion to still use WaitNamedPipe?

gdamore commented:
I think using WaitNamedPipe is probably more trouble than it's worth.  In
theory it would be better with the immediate wake up but as it doesn't
actually correct anything and as it blocks the caller I think it's better
not to use it and aim for simplicity.

5s was chosen to match TCP timeouts making this behave more like other
net.Pipe implementations. That said it is trivial to override as that is
just a default so I am happy with whatever you think is best there.

Keeping the old pipe open was intended to make it easier to distinguish
between the two cases of a listener not existing at all and a listener that
has not yet accepted the connection.  The former case you want to fail fast
without a time out.  The latter should keep trying because presumably
*something* is there and you have just missed it.

Again the semantic I am trying to provide is more like TCP or UNIX domain
pipes which both behave this way.

On Sat, Jul 14, 2018, 7:50 AM Olli Janatuinen <notifications@github.com>
wrote:

> @jstarks <https://github.com/jstarks> thanks for good comments.
>
> I changed default timeout to 250 ms and added client handle closing now on
> this commit: olljanat@972aaec
> <https://github.com/olljanat/go-winio/commit/972aaec17501edc2ae66b43541f799eec50cf7c5>
> and did some testing with Portainer and I can tell that this combination at
> least still fixes #67 <https://github.com/Microsoft/go-winio/issues/67>
>
> @gdamore <https://github.com/gdamore> I think that it is better if you
> comment about suggestion to still use WaitNamedPipe?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/go-winio/pull/80#issuecomment-405028207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPDfRMoGmzocewPA8PnfX8JKpKGOi12ks5uGgU1gaJpZM4U1QgD>
> .
>

olljanat commented:
@jstarks / @gdamore so what we can do get this one fixed?

There is already three PRs #75 , this one and #84 which all are trying to solve same issue but bit different way. We should choose which one to focus and do needed changes to get is merged.

gdamore commented:
Hmm.. I have a question.

If the client handle is closed, but we haven't called DisconnectNamedPipe nor closed the server handle, is the server pipe instance still retained and busy (so no new client handles will connect to it, and so that the we won't get ERROR_NOT_FOUND in CreateFile on the client?) 

If so, then closing the file handle would be perfectly reasonable.  In such a case, there would be no need to retain the client handle.  A better comment explaining this would be helpful.  There is still that race condition I mentioned, where some other clients connects before we do.  *That* would be unfortunate, and again would be easily fixed in a follow up.

jstarks commented:
What I've observed (and I think we have tests to confirm) is that keeping the server handle open is sufficient to retain ownership of the pipe and to ensure that clients get ERROR_PIPE_BUSY. So keeping the client handle open is not necessary.

gdamore commented:
Do you want me to follow up with a modification to this PR for the above alterations (2 second timeout, close client handle) or address at your end?

gdamore commented:
Ok, coming shortly.

gdamore commented:
Done.  Would be good to test it before integrating. :-)

jstarks commented:
Thanks! I'll take a look a little later today and merge it. If @olljanat has time to validate it in his workload, that would be useful.

olljanat commented:
@jstarks my workload looks working very nicely with this one.
@gdamore good work 👍

jstarks commented:
Thanks @gdamore for the fix, @olljanat for verification, and everyone for their patience on this.
